### PR TITLE
Fixes reward and text gathering in distributed training

### PR DESCRIPTION
**Pull request by edbeeching**
This PR fixes two issues:

- **Rewards** were being wrongly summed instead of gathered. While this did not affect the mean, other stats such as std and the distribution plots in wandb were affected.
- **Texts** were not being gathered across processes, leading to text only being reported from the main process.

I have tested in two settings [single node 8 GPU](https://wandb.ai/edbeeching/trl/runs/5ovy2waw/overview?workspace=user-edbeeching) and [single node 1 GPU](https://wandb.ai/edbeeching/trl/runs/6zz4yyh2?workspace=user-edbeeching). I launched both with `--ppo-config.mini-batch-size=1 --ppo-config.batch-size=2` to confirm the gather is working correctly.

This may impact #847

### CODE REVIEW

File: trl/trainer/ppo_trainer.py [hunk 1]
```diff
@@ -1310,17 +1320,21 @@ def log_stats(
...
+                    import torch.distributed as dist
 
-                table_rows = [list(r) for r in zip(*batch_list, rewards.cpu().tolist())]
-                logs.update({"game_log": wandb.Table(columns=[*columns_to_log, "reward"], rows=table_rows)})
-            # All reduce rewards if distributed
-            if self.is_distributed:
-                import torch.distributed as dist
+                    dist.barrier()
+                    gathered_batch_list = []
+                    for b in batch_list:
```
Comment (vwxyzjn) [inline trl/trainer/ppo_trainer.py:1328]: Similar as above.

File: trl/trainer/ppo_trainer.py [hunk 2]
```diff
@@ -1350,7 +1364,23 @@ def log_stats(
...
+                        raise ValueError(
+                            f"Columns to log {columns_to_log} are not present in the batch {batch.keys()}."
+                        )
+
+                    batch_list = [batch[column_to_log] for column_to_log in columns_to_log]
+
+                    dist.barrier()
+                    for b in batch_list:
+                        gather = [None for _ in range(dist.get_world_size())]
+                        dist.all_gather_object(gather, b)
```
Comment (vwxyzjn) [inline trl/trainer/ppo_trainer.py:1383]: What is this doing?

File: trl/trainer/ppo_trainer.py [hunk 3]
```diff
@@ -1297,6 +1297,16 @@ def log_stats(
...
 
+            # Gather rewards if distributed
+            if self.is_distributed:
+                import torch.distributed as dist
+
+                dist.barrier()
+                world_size = dist.get_world_size()
+                tensors = [torch.zeros_like(rewards) for _ in range(world_size)]
+                dist.all_gather(tensors, rewards)
+                rewards = torch.cat(tensors)
```
Comment (vwxyzjn) [inline trl/trainer/ppo_trainer.py:1308]: Maybe all these line of code can be replaced with self.accelerator.gather(rewards)


```suggestion
            # Gather rewards if distributed
            rewards = self.accelerator.gather(rewards).flatten()
```

**Review by vwxyzjn — COMMENTED** [review]
Very nice PR! Thanks @edbeeching this will make logging much better

File: trl/trainer/ppo_trainer.py (hunk 3 above)
Comment (edbeeching) [reply trl/trainer/ppo_trainer.py:1308]: Ah nice, I was not aware accelerate has this helper method, I will clean up.

File: trl/trainer/ppo_trainer.py (hunk 2 above)
Comment (edbeeching) [reply trl/trainer/ppo_trainer.py:1383]: Here I am sending the query & response from process n>1 to process 0. This is two way, perhaps there is a way to do this one way with send / recv? I don't often use those operations.

**Comment by lewtun** [comment]
Begone stale bot 🪄!

**Comment by edbeeching** [comment]
Yes, I will get back to this when I have the time.

**Comment by edbeeching** [comment]
@vwxyzjn I have rebased and updated the PR if you can take another look.

File: trl/trainer/ppo_trainer.py [hunk 4]
```diff
@@ -1368,16 +1368,6 @@ def log_stats(
...
-        else:
-            if self.is_distributed:
-                import torch.distributed as dist
-
-                if not isinstance(rewards, torch.Tensor):
-                    rewards = torch.tensor(rewards).to(self.current_device)
-
-                dist.barrier()
-                dist.all_reduce(rewards, op=torch.distributed.ReduceOp.SUM)
-
```
Comment (vwxyzjn) [inline trl/trainer/ppo_trainer.py:1380]: @lvwerra do these code do anything?

**Review by vwxyzjn — APPROVED** [review]
@edbeeching I have refactored just a bit :) There seems to be some unused code that I will have @lvwerra take a look at but otherwise LGTM.

Thanks a lot for the PR!

File: trl/trainer/ppo_trainer.py (hunk 4 above)
Comment (lvwerra) [reply trl/trainer/ppo_trainer.py:1380]: cc @younesbelkada :)

**Review by lvwerra — APPROVED** [review]
Looks good to me, I'll let @younesbelkada have a look too!

File: trl/trainer/ppo_trainer.py [hunk 5]
```diff
@@ -1336,17 +1337,16 @@ def log_stats(
                     raise ValueError(f"Columns to log {columns_to_log} are not present in the batch {batch.keys()}.")
 
                 batch_list = [batch[column_to_log] for column_to_log in columns_to_log]
+                if self.is_distributed:
+                    self.accelerator.wait_for_everyone()
+                    gathered_batch_list = []
+                    for b in batch_list:
```
Comment (lewtun) [inline trl/trainer/ppo_trainer.py:1343]: turbo nit to follow HF conventions to use descriptive names
```suggestion
                    for batch in batch_list:
```

**Review by lewtun — APPROVED** [review]
LGTM with a turbo nit about variable names :)

I also have the same question as @vwxyzjn about the remove of some distributed code, but @younesbelkada is the best person to answer that

**Comment by lvwerra** [comment]
Is this ready to be merged?

**Comment by vwxyzjn** [comment]
Merged! Thanks @edbeeching @lewtun @lvwerra !
